### PR TITLE
Add week, week_of_year Presto functions

### PR DESCRIPTION
**Pull request by pramodsatya**
Add Presto functions week and week_of_year (alias of week).

### CODE REVIEW

**Comment by majetideepak** [comment]
Add documentation here https://facebookincubator.github.io/velox/functions/datetime.html. Can copy the description from 
https://prestodb.io/docs/current/functions/datetime.html

File: velox/functions/prestosql/DateTimeFunctions.h [hunk 1]
```diff
@@ -131,6 +131,37 @@ struct TimestampWithTimezoneSupport {
 
 } // namespace
 
+template <typename T>
+struct WeekFunction : public InitSessionTimezone<T>,
+                      public TimestampWithTimezoneSupport<T> {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE int64_t getWeek(const std::tm& time) {
+    int daysPerWeek = 7;
```
Comment (majetideepak) [inline velox/functions/prestosql/DateTimeFunctions.h:140]: `const int`

File: velox/functions/prestosql/registration/DateTimeFunctionsRegistration.cpp [hunk 2]
```diff
@@ -60,6 +63,10 @@ void registerSimpleFunctions() {
   registerFunction<YearOfWeekFunction, int64_t, Date>({"yow", "year_of_week"});
   registerFunction<YearOfWeekFunction, int64_t, TimestampWithTimezone>(
       {"yow", "year_of_week"});
+  registerFunction<WeekFunction, int64_t, Timestamp>({"week_of_year"});
+  registerFunction<WeekFunction, int64_t, Date>({"week_of_year"});
+  registerFunction<WeekFunction, int64_t, TimestampWithTimezone>(
+      {"week_of_year"});
```
Comment (majetideepak) [inline velox/functions/prestosql/registration/DateTimeFunctionsRegistration.cpp:69]: is there a short name `woy` similar to `yow` above?

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 3]
```diff
@@ -377,6 +377,110 @@ TEST_F(DateTimeFunctionsTest, yearTimestampWithTimezone) {
...
+  EXPECT_EQ(40, week(Timestamp(4000000000, 123000000)));
+  EXPECT_EQ(34, week(Timestamp(998474645, 321000000)));
+}
+
+TEST_F(DateTimeFunctionsTest, weekDate) {
+  const auto week = [&](std::optional<Date> date) {
+    return evaluateOnce<int64_t>("week(c0)", date);
+  };
+
+  EXPECT_EQ(std::nullopt, week(std::nullopt));
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:404]: Please add something like 

auto parseDate = [](const std::string& dateStr) {
    Date returnDate;
    parseTo(dateStr, returnDate);
    return returnDate;
  };

and use the string representation of the Date for the tests to be more readable. e.g.
EXPECT_EQ(0, week(parseDate("1970-01-01")));

Same for the others if possible.

**Review by aditi-pandit — COMMENTED** [review]
Thanks Pramod. Few comments.

File: velox/functions/prestosql/DateTimeFunctions.h [hunk 4]
```diff
@@ -131,6 +131,37 @@ struct TimestampWithTimezoneSupport {
...
+    int64_t week = (time.tm_yday + daysPerWeek -
+                    (time.tm_wday ? (time.tm_wday - 1) : (daysPerWeek - 1))) /
```
Comment (majetideepak) [inline velox/functions/prestosql/DateTimeFunctions.h:142]: `daysPerWeek -
                    (time.tm_wday ? (time.tm_wday - 1) : (daysPerWeek - 1)` is hard to follow.
https://cplusplus.com/reference/ctime/tm/ says `tm_wday` is days since Sunday. Doesn't the above equation decrease as the value of `tm_wday` increases?

File: velox/functions/prestosql/registration/DateTimeFunctionsRegistration.cpp (hunk 2 above)
Comment (pramodsatya) [reply velox/functions/prestosql/registration/DateTimeFunctionsRegistration.cpp:69]: Unfortunately not, Presto documentation (https://prestodb.io/docs/current/functions/datetime.html#week) doesn't have any such function.

File: velox/functions/prestosql/DateTimeFunctions.h (hunk 4 above)
Comment (aditi-pandit) [reply velox/functions/prestosql/DateTimeFunctions.h:142]: Interesting point. Have you looked at the Presto code to calculate the same ?

https://en.wikipedia.org/wiki/ISO_week_date also has algorithms for calculating week of the year.

Comment (pramodsatya) [reply velox/functions/prestosql/DateTimeFunctions.h:142]: Yes actually it is difficult to follow, I had a separate case for leap years earlier but then came across this implementation used in FreeBSD here: https://cplusplus.com/forum/general/220155/ . This seemed simpler and it worked correctly for edge cases so I followed the same. From what I understood, the above equation ensures the week with first Thursday is counted as week 1 by adding ``daysPerWeek - (time.tm_wday - 1)``. I will follow the algorithm on Wikipedia shared by Aditi since it is more clear, thanks!

Comment (pramodsatya) [reply velox/functions/prestosql/DateTimeFunctions.h:142]: > 

Thanks for sharing, will follow the same

Comment (aditi-pandit) [reply velox/functions/prestosql/DateTimeFunctions.h:142]: In any case, what is the Presto java calculation ? We should be consistent with that in general.

File: velox/functions/prestosql/registration/DateTimeFunctionsRegistration.cpp [hunk 5]
```diff
@@ -30,6 +30,9 @@ void registerSimpleFunctions() {
   registerFunction<YearFunction, int64_t, Timestamp>({"year"});
   registerFunction<YearFunction, int64_t, Date>({"year"});
   registerFunction<YearFunction, int64_t, TimestampWithTimezone>({"year"});
+  registerFunction<WeekFunction, int64_t, Timestamp>({"week"});
```
[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/functions/prestosql/registration/DateTimeFunctionsRegistration.cpp]: Thanks for sharing, changed accordingly

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 6]
```diff
@@ -404,6 +404,128 @@ TEST_F(DateTimeFunctionsTest, yearTimestampWithTimezone) {
...
+  EXPECT_EQ(1, week(Timestamp(0, 0)));
+  EXPECT_EQ(53, week(Timestamp(-1, 9000)));
+  EXPECT_EQ(40, week(Timestamp(4000000000, 123000000)));
+  EXPECT_EQ(34, week(Timestamp(998474645, 321000000)));
+
+  setQueryTimeZone("Pacific/Apia");
+  EXPECT_EQ(std::nullopt, week(std::nullopt));
+  EXPECT_EQ(53, week(Timestamp(0, 0)));
+  EXPECT_EQ(53, week(Timestamp(-1, 12300000000)));
+  EXPECT_EQ(40, week(Timestamp(4000000000, 123000000)));
```
Comment (majetideepak) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:422]: do we have a `parseTimestamp` like `parseDate`?

Comment (pramodsatya) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp]: Thanks for the suggestion, the function ```fromTimeString``` takes input in HH:MM:SS format and is more readable.

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 7]
```diff
@@ -485,6 +485,128 @@ TEST_F(DateTimeFunctionsTest, yearTimestampWithTimezone) {
...
+        "week(c0)",
+        std::make_optional(Timestamp(util::fromTimeString(date), 0)));
+  };
+
+  EXPECT_EQ(1, week("00:00:00"));
+  EXPECT_EQ(49, week("11:59:59"));
+  EXPECT_EQ(22, week("06:01:01"));
+  EXPECT_EQ(52, week("12:00:01"));
+
+  setQueryTimeZone("Pacific/Apia");
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:500]: Should this be Asia and not Apia ?

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 8]
```diff
@@ -485,6 +485,128 @@ TEST_F(DateTimeFunctionsTest, yearTimestampWithTimezone) {
           "year(c0)", std::nullopt, std::nullopt));
 }
 
+TEST_F(DateTimeFunctionsTest, week) {
+  const auto week = [&](StringView date) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:489]: This seems a time string and not date. Please change the name of the parameter. 

Also should the parameter be const StringView&

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 9]
```diff
@@ -485,6 +485,128 @@ TEST_F(DateTimeFunctionsTest, yearTimestampWithTimezone) {
...
+    return evaluateOnce<int64_t>(
+        "week(c0)",
+        std::make_optional(Timestamp(util::fromTimeString(date), 0)));
+  };
+
+  EXPECT_EQ(1, week("00:00:00"));
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:495]: I'm a bit lost about how a time string can have a week. I would've expected this to error. Please can you explain the behavior of these tests.

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 10]
```diff
@@ -485,6 +485,128 @@ TEST_F(DateTimeFunctionsTest, yearTimestampWithTimezone) {
...
+  const auto week = [&](std::optional<Date> date) {
+    return evaluateOnce<int64_t>("week(c0)", date);
+  };
+  auto parseDate = [&](const std::string& dateStr) {
+    Date returnDate;
+    parseTo(dateStr, returnDate);
+    return returnDate;
+  };
+
+  EXPECT_EQ(53, week(parseDate("1919-12-31")));
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:517]: Please try some corner cases like a really small date like "0001 - 01- 01" and very large dates like "9999-12-31"

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 11]
```diff
@@ -485,6 +485,128 @@ TEST_F(DateTimeFunctionsTest, yearTimestampWithTimezone) {
...
+      5,
+      evaluateWithTimestampWithTimezone<int64_t>(
+          "week(c0)", -123456789000, "+03:00"));
+  EXPECT_EQ(
+      std::nullopt,
+      evaluateWithTimestampWithTimezone<int64_t>(
+          "week(c0)", std::nullopt, std::nullopt));
+}
+
+TEST_F(DateTimeFunctionsTest, weekOfYear) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:547]: The same comments in the week test apply here. Please can you check again.

**Review by aditi-pandit — COMMENTED** [review]
Thanks for adding this function ! Have a few small comments.

File: velox/functions/prestosql/DateTimeFunctions.h [hunk 12]
```diff
@@ -133,6 +135,59 @@ struct TimestampWithTimezoneSupport {
...
+
+  FOLLY_ALWAYS_INLINE int64_t getWeek(const std::tm& time) {
+    // The computation of ISO week from date follows the algorithm here:
+    // https://en.wikipedia.org/wiki/ISO_week_date
+    int64_t week = floor(
+                       10 + (time.tm_yday + 1) -
+                       (time.tm_wday ? time.tm_wday : kDaysInWeek)) /
+        kDaysInWeek;
+
+    if (week == 0) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/DateTimeFunctions.h:151]: Please add the comments like those from here https://cplusplus.com/forum/general/220155/ as it makes the code more understandable.

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 13]
```diff
@@ -485,6 +485,119 @@ TEST_F(DateTimeFunctionsTest, yearTimestampWithTimezone) {
           "year(c0)", std::nullopt, std::nullopt));
 }
 
+TEST_F(DateTimeFunctionsTest, weekDate) {
+  const auto week = [&](std::optional<Date> date) {
+    return evaluateOnce<int64_t>("week(c0)", date);
+  };
+  auto parseDate = [&](const StringView& dateStr) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:492]: This lambda repeats multiple time. Consider promoting it to a method of DateTimeFunctionsTest to avoid code duplication.

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 14]
```diff
@@ -485,6 +485,119 @@ TEST_F(DateTimeFunctionsTest, yearTimestampWithTimezone) {
...
+    Date returnDate;
+    parseTo(dateStr, returnDate);
+    return returnDate;
+  };
+
+  EXPECT_EQ(1, week(parseDate("1919-12-31")));
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:498]: consider refactoring to allow for shorter code: `EXPECT_EQ(1, week("1919-12-31"));`

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 15]
```diff
@@ -485,6 +485,119 @@ TEST_F(DateTimeFunctionsTest, yearTimestampWithTimezone) {
...
+  EXPECT_EQ(53, week(parseDate("1960-01-01")));
+  EXPECT_EQ(53, week(parseDate("1960-01-03")));
+  EXPECT_EQ(1, week(parseDate("1960-01-04")));
+  EXPECT_EQ(1, week(parseDate("1969-12-31")));
+  EXPECT_EQ(1, week(parseDate("1970-01-01")));
+  EXPECT_EQ(1, week(parseDate("0001-01-01")));
+  EXPECT_EQ(52, week(parseDate("9999-12-31")));
+}
+
+TEST_F(DateTimeFunctionsTest, week) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:511]: week -> weekTimestamp for clarity

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 16]
```diff
@@ -485,6 +485,119 @@ TEST_F(DateTimeFunctionsTest, yearTimestampWithTimezone) {
...
+  EXPECT_EQ(1, week("00:00:00", "+12:00"));
+  EXPECT_EQ(47, week("11:59:59", "-12:00"));
+  EXPECT_EQ(47, week("11:59:59", "+12:00"));
+  EXPECT_EQ(33, week("06:01:01", "-12:00"));
+  EXPECT_EQ(34, week("06:01:01", "+12:00"));
+  EXPECT_EQ(47, week("12:00:01", "-12:00"));
+  EXPECT_EQ(47, week("12:00:01", "+12:00"));
+}
+
+TEST_F(DateTimeFunctionsTest, weekOfYearDate) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:544]: Since week_of_year is an alias of the "week" consider dropping these test cases or refactor to avoid copy-paste.

File: velox/functions/prestosql/DateTimeFunctions.h (hunk 12 above)
Comment (pramodsatya) [reply velox/functions/prestosql/DateTimeFunctions.h:151]: Updated, thanks for the suggestion!

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp (hunk 13 above)
Comment (pramodsatya) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp]: Done, thanks!

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp (hunk 16 above)
Comment (pramodsatya) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp]: Thanks for the suggestion, used bitwise AND between the results of ```week``` and ```week_of_year``` in the week lambdas to ensure both return the same expected value.

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 17]
```diff
@@ -485,6 +491,73 @@ TEST_F(DateTimeFunctionsTest, yearTimestampWithTimezone) {
           "year(c0)", std::nullopt, std::nullopt));
 }
 
+TEST_F(DateTimeFunctionsTest, weekDate) {
+  const auto week = [&](const StringView& date) {
+    return evaluateOnce<int64_t>(
+               "week(c0)", std::make_optional(parseDate(date)))
+               .value() &
```
[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:498]: Apologies for missing this corner case, I used ```&``` instead of ```^```, will update accordingly.

[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:498]: Let's not use `^`. This makes the code very hard to understand and is easy to make a mistake again.

[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp]: Got it, thanks! ```ASSERT_EQ``` resulted in compilation error due to void and optional<long long> being returned by different paths in the function, so ```VELOX_CHECK``` is used instead.

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 18]
```diff
@@ -485,6 +491,85 @@ TEST_F(DateTimeFunctionsTest, yearTimestampWithTimezone) {
...
+TEST_F(DateTimeFunctionsTest, weekDate) {
+  const auto weekDate = [&](const StringView& date) {
+    auto week =
+        evaluateOnce<int64_t>("week(c0)", std::make_optional(parseDate(date)))
+            .value();
+    auto weekOfYear =
+        evaluateOnce<int64_t>(
+            "week_of_year(c0)", std::make_optional(parseDate(date)))
+            .value();
+    VELOX_CHECK(
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:503]: Can we use EXPECT_EQ or VELOX_CHECK_EQ to get a better error message, i.e. have the values being compared be included in the message?

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 19]
```diff
@@ -485,6 +491,85 @@ TEST_F(DateTimeFunctionsTest, yearTimestampWithTimezone) {
           "year(c0)", std::nullopt, std::nullopt));
 }
 
+TEST_F(DateTimeFunctionsTest, weekDate) {
+  const auto weekDate = [&](const StringView& date) {
+    auto week =
+        evaluateOnce<int64_t>("week(c0)", std::make_optional(parseDate(date)))
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:497]: > std::make_optional(parseDate(date))

This code is rather long and repeats. Consider extracting a temp variable. Ditto other tests.

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 20]
```diff
@@ -485,6 +491,85 @@ TEST_F(DateTimeFunctionsTest, yearTimestampWithTimezone) {
...
+  EXPECT_EQ(1, weekTimestamp("00:00:00"));
+  EXPECT_EQ(10, weekTimestamp("11:59:59"));
+  EXPECT_EQ(51, weekTimestamp("06:01:01"));
+  EXPECT_EQ(24, weekTimestamp("06:59:59"));
+  EXPECT_EQ(27, weekTimestamp("12:00:01"));
+  EXPECT_EQ(7, weekTimestamp("12:59:59"));
+}
+
+TEST_F(DateTimeFunctionsTest, weekTimestampWithTimezone) {
+  const auto weekTimestampTimezone = [&](const StringView& time,
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:548]: StringView's can be passed by value, but here it is confusing to use Velox's StringView. Consider changing to const char* or std::string_view.

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp (hunk 19 above)
Comment (pramodsatya) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:497]: Done, thanks!

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp (hunk 18 above)
Comment (pramodsatya) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:503]: Changed to VELOX_CHECK_EQ, thanks for the suggestion!

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 21]
```diff
@@ -223,6 +223,12 @@ class DateTimeFunctionsTest : public functions::test::FunctionBaseTest {
     }
     return signatureStrings;
   }
+
+  Date parseDate(const StringView& dateStr) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:227]: I believe there are existing tests that use this as a lambda. We can delete these lambdas now, right?

**Review by mbasmanova — APPROVED** [review]
@pramodsatya Thanks.
